### ISSUE TITLE
calculating map

### ISSUE THREAD

**Issue by Jin-S-Kim** [AlexeyAB/darknet#2253]
I train a new object(custom data) by using yolo_mark and dartknet and get new weights. And I want to see the result of training so I use './darknet detector map ~data ~cfg ~weights'.

If set repository of valid to images that used for training(these image files have txt files). 
![1](https://user-images.githubusercontent.com/31817047/51442020-9b663500-1d1b-11e9-8fea-142d1c11bb96.png)

It calculate map.
And If I set repository of valid to images that for test(these images files have no txt files) it cannot calculate map and this is the result 
![2019-01-21 01-22-22](https://user-images.githubusercontent.com/31817047/51442010-7f629380-1d1b-11e9-99a1-d8a1382e4950.png)

So I make txt file of test images using new weights(that I get for training by training images) and put txt files in directory where test images is in. 

![1](https://user-images.githubusercontent.com/31817047/51442045-d4060e80-1d1b-11e9-88b9-6d5d542f0b1f.png)

And it calculate map.

![123](https://user-images.githubusercontent.com/31817047/51442048-dbc5b300-1d1b-11e9-8a7c-fe9b1349969b.png)

Does the map have to be a class number of the object and a txt file containing information from the bounding box? Then, if I make a map by making a txt file of test images with the weight I trained, the detection result will be a ground trough. Shouldn't the map be 100 percent?
How can this command calculate map?

**Comment by AlexeyAB** [comment]
> And If I set repository of valid to images that for test(these images files have no txt files) it cannot calculate map and this is the result

Both Training, Validation and Test datasets must have txt-labels files with ground-truth boxes.

----

> So I make txt file of test images using new weights(that I get for training by training images) and put txt files in directory where test images is in.

Did you do these files by using pseudo-labeling using `-save_labels` flag?
`./darknet detector test data/obj.data yolov3_obj.cfg backup/yolov3_obj_10000.weights -save_labels < valid.txt`

----

> Shouldn't the map be 100 percent?

No, it shouldn't. Because your labels are generated for the default Confidence-threshold `-thresh 0.25`, 
but mAP is calculated for all possible Confidence-thresholds

**Comment by Sudhakar17** [comment]
@AlexeyAB 
I checked the function validate_detector_map. There are two variables thresh_calc_avg_iou = 0.25 and iou_thresh = 0.5.  I understand the second one which is compared with cur_iou of the detection. I don't understand the first one which is compared with prob. Can you please tell me why it is like that?

**Comment by AlexeyAB** [comment]
@Sudhakar17 

`thresh_calc_avg_iou = 0.25` - is a confidence-threshold (for probability), it is used only for output F1, TN/FP/TP, Precision, Recall, IoU for the specified confidence threshold

`iou_thresh = 0.5` is a IoU-threshold, that is used for mAP calculation, it calculates `mAP@0.5` by default (iou_thresh = 0.5)

**Comment by Sudhakar17** [comment]
@AlexeyAB 
Thanks a lot. I have one more question related to this one. I used test_detector to save the bbox co-ordinates, class prediction probability (-ext_output) in the text file. Then I used some python scripts to calculate mAP from the predicted results. 

I am getting 54.37 % mAP for COCO-Val set.(C implementation) and 45.74% mAP for the same validation set (Python implementaion)

you calculated the results from validate_detector_map and test_detector(inference) functions are the same, right?

**Comment by Sammytop123** [comment]
Hey Alex,
The `./darknet detector map` command takes my images and feeds them into the network, gets their predictions and compares them to the ground truth labels to produce the AP and mAP statistics.

For a project, I'm trying to produce the AP and mAP statistics in the same way as this script with the adjustment that I have already my predictions for every img, meaning I got my ground_truth labels and the predicted labels for every img. How can I use your script to calculate the AP and mAP stats ?

Thx
